app/models/relations/scopes/visible.rb

```diff
@@ -34,10 +34,25 @@ module Visible
       # Returns all relationships visible to the user. The relationships have to be:
       #   * Start (from_id) on a work package visible to the user (view_work_packages in the work package's project)
       #   * End (to_id) on a work package visible to the user (view_work_packages in the work package's project)
+      #
+      # In some cases, the resulting SQL query as is might not run in a performant matter. This happens in
+      # cases where there are a lot of work packages and relations. PostgreSql then sometimes chooses to use a
+      # full table scan on work_packages expecting a lot of results.
+      # Then, the query can be optimized by providing a +work_package_focus_scope+ which is used as a subquery
+      # on work_packages on the id column. This is beneficial if not all work packages are to be
+      # considered but only a subset. As this directly excludes work packages, it can also be used
+      # the wrong way.
       # @param [User] user
-      def visible(user = User.current)
-        where(from_id: WorkPackage.visible(user))
-          .where(to_id: WorkPackage.visible(user))
+      # @param [ActiveRecord::Relation, Arel::SelectManager] work_package_focus_scope
+      def visible(user = User.current, work_package_focus_scope: nil)
+        wp_arel = work_package_focus_scope.respond_to?(:arel) ? work_package_focus_scope.arel : work_package_focus_scope
+        visible_work_packages = WorkPackage.visible(user)
+
+        visible_work_packages = visible_work_packages.where(WorkPackage.arel_table[:id].in(wp_arel)) if wp_arel
```

Comment:
```suggestion
        visible_work_packages = WorkPackage.visible(user)

        if work_package_focus_scope
          wp_arel = work_package_focus_scope.respond_to?(:arel) ? work_package_focus_scope.arel : work_package_focus_scope
          visible_work_packages = visible_work_packages.where(WorkPackage.arel_table[:id].in(wp_arel))
        end
```

Reply:
I tried this briefly but rubocop complained about complexity, so I abandoned it again.

Reply:
I just wanted to make it clearer that `wp_arel` is needed only if `work_package_focus_scope` is there

```suggestion
        visible_work_packages = WorkPackage.visible(user)

        wp_arel = work_package_focus_scope.respond_to?(:arel) ? work_package_focus_scope.arel : work_package_focus_scope
        visible_work_packages = visible_work_packages.where(WorkPackage.arel_table[:id].in(wp_arel)) if wp_arel
```